### PR TITLE
Horreum UI refactoring

### PR DESCRIPTION
**Pull request by johnaohara**
Horreum UI refactoring.

In summary, here the most important changes:
- [x] Main toolbar refactored as left-sidebar
- [x] Test-specific toolbar improved
   - [x] Runs and datasets showed as inner tabs
   - [x] Changes and reports are no more standalone pages, they are under a specific test
   - [x] "General" tab renamed as "Settings"
- [x] Folders navigation refactored using `breadcrumb`

### CODE REVIEW

**Comment by willr3** [comment]
Screenshot from 2024-02-07 21-04-33

Overall I like the idea of moving items under the associated `Test` and preserving the selected test when navigating between them.

My opinion on changes:
* The `login` button should stay right aligned in the top nav.
* Only 2 (3 with `Admin`) options in the sidebar uses a lot of vertical space. 
* What happened to `Runs`? Users upload `Runs` not `Datasets` so I think `Datasets` are what should go away, not `Runs`.

**Comment by johnaohara** [comment]
>    The login button should stay right aligned in the top 

That was my intention, it is layout issue that needs resolving

>    Only 2 (3 with Admin) options in the sidebar uses a lot of vertical space.

I was thinking about being about to "favourite" tests which would appear under the "Test" menu, this would fill this sidebar up.

It is also possible to collapse the sidebar so you get the full width of the screen by pressing the  button
image

The other option is to put `Tests` / `Schema` / `Admin` back on the top menu and some of the 

>    What happened to Runs? Users upload Runs not Datasets so I think Datasets are what should go away, not Runs.

I was experimenting with simplifying the UI, atm the core abstraction for change detection is dataset. The "Runs" View allows you to compare runs and delete them, but is similar to datasets.

Overall we need to think about what is the best way to visualize the Run/Dataset abstraction, or if we need that abstraction at all

**Comment by johnaohara** [comment]
The "Folder" navigation could move into the navigation bar
image

and we get a lot of space back at the top of the main view

image

instead of what we have right now:

image

**Comment by lampajr** [comment]
## What I did
In the previous commits I have fixed (hopefully) some UI issues that I'll summarize here:
1. Aligned the header toolbar to the right such that the `Log in` is in the top-right corner
image
2. Simplified the `changes` tab query params, removing the weird `http://localhost:8080/test/113/?test=[object%20Object]` test param (which is not really needed as already present in the URI)
image
3. Fixed the Alert box message as it was overflowing when the left sidebar was opened
4. Fixed the reports navigation as it was commented out
   * Here I re-established the same UI that was there before, I simply changed the full URI to contain the `test/:testId` as root.
   * For instance, I changed from `/reports/table/config/:configId` to `/test/:testId/reports/table/config/:configId`
   * Now the `reports` link moves the the user back to the reports tab for that specifc `test`
image

## Open points
In conclusion I think that from this pull request point of view we have still ~2 open points:
* Is the `reports` navigation I re-established fine? Or were we thinking about a different approach?
* Right now the folder selection is just a dropdown list
image
   * Once selected the folder name is not shown in that dropdown button
image
   * Do we want to implement the folder structure as mentioned/showed in this comment https://github.com/Hyperfoil/Horreum/pull/1290#issuecomment-1933469772 ?

**Comment by lampajr** [comment]
Alternatively another option for the Folder structure could be to use the `breadcrumb` approach like:
image
image
image

I still don't know where the `Team Selector` could be placed, as I don't really like it next to the breadcrumb.
Any opinion on this approach?

File: horreum-web/src/components/SavedTabs.tsx [hunk 1]
```diff
@@ -101,7 +101,7 @@ export default function SavedTabs(props: SavedTabsProps) {
                             const childProps = children[activeKey.current].props
                             if ("onSave" in childProps) {
                                 childProps
-                                    .onSave()
+                                    .onSave?.()
```
Comment (lampajr) [inline horreum-web/src/components/SavedTabs.tsx:104]: I don't think this is actually required as `onSave` is a mandatory property, isn't it?
```suggestion
                                    .onSave()
```

File: horreum-web/src/domain/alerting/Changes.tsx [hunk 2]
```diff
@@ -204,16 +205,21 @@ export const flattenNode = (arr : Array<any> | undefined) => {
...
     const navigate = useNavigate()
     const params = new URLSearchParams(location.search)
-    // eslint-disable-next-line
-    const paramTest = useMemo(() => params.get("test") || undefined, [])
-    const paramFingerprint = params.get("fingerprint")
     const teams = useSelector(teamsSelector)
-    const [selectedTest, setSelectedTest] = useState<SelectedTest>()
+    const newTest = { id: props.testID } as SelectedTest;
+    // Do we need this? the test id is provided by props therefore it should not change
+    const [selectedTest, setSelectedTest] = useState<SelectedTest>(newTest)
```
Comment (lampajr) [inline horreum-web/src/domain/alerting/Changes.tsx:220]: Do we need this? the test id is provided by props therefore it should not change. In fact the function `setSelectedTabs` is not called anywhere

File: horreum-web/src/domain/tests/AllTests.tsx [hunk 3]
```diff
@@ -394,22 +372,20 @@ export default function AllTests() {
...
+
     const loadTests = () => {
-        setLoading(true)
         fetchTestsSummariesByFolder(alerting, SortDirection.Ascending, pagination.folder, pagination.limit, pagination.page, rolesFilter.key, )
-            .then(summary => {setTests(summary.tests?.map(t => mapTestSummaryToTest(t)) || [])
-                if (summary.count) {
-                    setCount(summary.count)
-                }
-            })
-            .finally( () => setLoading(false))
```
Comment (lampajr) [inline horreum-web/src/domain/tests/AllTests.tsx]: I see the `loading` management has been removed, this means that `const [loading, setLoading] = useState(false)` would be no longer needed as no more used anywhere here. Was that intentional?

File: horreum-web/src/domain/tests/AllTests.tsx [hunk 4]
```diff
@@ -394,22 +372,25 @@ export default function AllTests() {
...
+
     const loadTests = () => {
-        setLoading(true)
         fetchTestsSummariesByFolder(alerting, SortDirection.Ascending, pagination.folder, pagination.limit, pagination.page, rolesFilter.key, )
-            .then(summary => {setTests(summary.tests?.map(t => mapTestSummaryToTest(t)) || [])
+            .then(summary => {
+                setTests(summary.tests?.map(t => mapTestSummaryToTest(t)) || [])
                 if (summary.count) {
                     setCount(summary.count)
                 }
```
Comment (lampajr) [inline horreum-web/src/domain/tests/AllTests.tsx:391]: With commit https://github.com/Hyperfoil/Horreum/pull/1290/commits/9781289787ac737f8b749e60680db44e9b1c1a1f I had to re-add this `setCount` call otherwise the pagination would not work as expected (always set to 0)

**Comment by lampajr** [comment]
As discussed added additional commit (https://github.com/Hyperfoil/Horreum/pull/1290/commits/ca3fff3452e872dc0df7da8a40de8db0cf3b96e2) to add folders navigation using breadcrumb:
image

I tried to keep the folder structure, i.e., `Horreum` as root of the tree and then all the folders, an alternative could be to remove that root layer:
image

**Comment by lampajr** [comment]
For better navigation the last commit introduce the original folder where the test resides, e.g.,
image

**Comment by lampajr** [comment]
Since we changed some of the URL paths, e.g., for `changes`, IIUC we should change the URLs we inject into the email notification as well (assuring they will point to the correct resources)

**Comment by johnaohara** [comment]
> 
> I tried to keep the folder structure, i.e., `Horreum` as root of the tree and then all the folders, an alternative could be to remove that root layer: image

I think the drop down with "24" is confusing, it is not obvious what the "24" refers to.  I think using "Test" in the breadcrumb for the dropdown might be more intuitive for users;

Screenshot from 2024-04-09 05-57-05

I would also drop the "<" arrow from Horreum, that way it appears to be a tree, with "Horreum" as the root

**Comment by johnaohara** [comment]
When I click on a "Run" of a "Dataset", the breadcrumb disappears.

I think we should retain the breadcrumb on all views

Screenshot from 2024-04-09 06-03-00

**Comment by johnaohara** [comment]
> Is the `reports` navigation I re-established fine? Or were you thinking about a different approach?
> 
>     * Maybe this is a question more for @johnaohara

@lampajr  What changes were made to the `reports` navigation? I can not see an issue, but was wondering what had changed?

File: horreum-web/src/domain/alerting/Changes.tsx (hunk 2 above)
Comment (johnaohara) [reply horreum-web/src/domain/alerting/Changes.tsx:220]: I think we can remove this now

**Comment by lampajr** [comment]
> > Is the `reports` navigation I re-established fine? Or were you thinking about a different approach?
> > ```
> > * Maybe this is a question more for @johnaohara
> > ```
> 
> @lampajr What changes were made to the `reports` navigation? I can not see an issue, but was wondering what had changed?

@johnaohara actually nothing changed (it was simply commented out, so I re-established it), the only difference with respect to the `as-is` is the URL that hanged from `/reports/table/config/:configId` to `/test/:testId/reports/table/config/:configId`

> I think the drop down with "24" is confusing, it is not obvious what the "24" refers to. I think using "Test" in the breadcrumb for the dropdown might be more intuitive for users;

I agree, much cleaner and easier to underastand. Working on it.

> When I click on a "Run" of a "Dataset", the breadcrumb disappears.
I think we should retain the breadcrumb on all views

Oh, good catch! I did not notice. Working on this as well.

**Comment by lampajr** [comment]
I was able to keep the breadcrumb on `run` and `dataset` view but unfortunately we don't have information about the folder (this could require some refactoring to pass that information down till the final view).

This is the actual result
image

Removed the number of folders from the dropdown, this is the result as you already showed:
image
image
image

File: horreum-web/src/components/SavedTabs.tsx (hunk 1 above)
Comment (johnaohara) [reply horreum-web/src/components/SavedTabs.tsx:104]: It has been a while since this change was made, but... iirc, i **think** I changed `SavedTabs.tsx` for the "Save" button to be optional, or at least that was the intention.  Not all tabs had a natural state that should be saved when navigating away from them, or I had moved some tabs that handling saving state slightly differently.

File: horreum-web/src/domain/tests/AllTests.tsx (hunk 4 above)
Comment (johnaohara) [reply horreum-web/src/domain/tests/AllTests.tsx:398]: ok

File: horreum-web/src/domain/tests/AllTests.tsx (hunk 3 above)
Comment (johnaohara) [reply horreum-web/src/domain/tests/AllTests.tsx]: @lampajr is this still an question? the code is flagged as "outdated"

Comment (lampajr) [reply horreum-web/src/domain/tests/AllTests.tsx]: yeah it is still valid question as the `loading` state is no more managed, therefore I was wondering if that has been done on purpose or not

**Comment by johnaohara** [comment]
On the `Tests` view, there is the option to filter tests by ownership in a dropdown.

That same dropdown is in the Schema view, but is not currently functional.  I think we should get this wired up

Screenshot from 2024-04-12 06-15-14

File: horreum-web/src/domain/tests/AllTests.tsx (hunk 3 above)
Comment (lampajr) [reply horreum-web/src/domain/tests/AllTests.tsx]: yeah sure

File: horreum-web/src/components/SavedTabs.tsx (hunk 1 above)
Comment (lampajr) [reply horreum-web/src/components/SavedTabs.tsx:104]: yeah makes sense, I checked further and the props could be either `FragmentTabProps` or `SavedTabProps` and the `onSave` method is present only in the latter (where it must be always present). 
Moreover you are checking a couple of lines before if `onSave` is present in the child props so I think it is save to remove that `?.` notation. Anyway just a minor :)

Comment (johnaohara) [reply horreum-web/src/components/SavedTabs.tsx:104]: sure, change it to `.onSave()`

File: horreum-web/src/domain/tests/AllTests.tsx (hunk 3 above)
Comment (lampajr) [reply horreum-web/src/domain/tests/AllTests.tsx]: Done

**Comment by lampajr** [comment]
Hi @johnaohara 

I addressed the other minors, especially I re-instated the `setLoading` on `AllTests` page.

> That same dropdown is in the Schema view, but is not currently functional. I think we should get this wired up

Yeah that's something we should do, I checked and it seems this is not supported from the backend therefore it is not about UI changes only. Do you want to do that as part of this PR? Or do you prefer to create a separate issue to manage both backend and frontend?

**Comment by lampajr** [comment]
> If it is not supported in the backend, we should hide the component in the UI and add that in a separate PR

I agree, hidden in this pr and created https://github.com/Hyperfoil/Horreum/issues/1589 for its implementation.
I raised that as bug since it is actually showed in production so users would exepct to work.

**Comment by johnaohara** [comment]
@lampajr please can you change title the PR, squash commits and remove from Draft status?

**Comment by shivam-sharma7** [comment]
@johnaohara will we nee all commentend code, I am asking because its kind of mesh.

**Comment by johnaohara** [comment]
> @johnaohara will we nee all commentend code, I am asking because its kind of mesh.

Not in this instance.  In this PR, the UI is being re-organised as an interim step, but there is more re factoring to be performed to allow for Unit testing etc.  As we perform work to improve the UI components, rather than just moving them around, I think we want more details about the components then

**Comment by whitingjr** [comment]
Introducing 3 clicks to navigation is a backward step imo.
1. click menu to open
2. click Test or Schema
3. click menu to hide
Created #1622 for a feature request to make it use mouse events to open/close.
